src/scales/scale.time.js

```diff
@@ -625,21 +625,46 @@ module.exports = function(Chart) {
 			return ticksFromTimestamps(ticks, me._majorUnit);
 		},
 
-		getLabelForIndex: function(index, datasetIndex) {
+		getLabelForIndex: function(index) {
 			var me = this;
 			var data = me.chart.data;
 			var timeOpts = me.options.time;
 			var label = data.labels && index < data.labels.length ? data.labels[index] : '';
-			var value = data.datasets[datasetIndex].data[index];
+			var len = data.labels.length;
+			var i, momentDate;
 
-			if (helpers.isObject(value)) {
-				label = me.getRightValue(value);
-			}
 			if (timeOpts.tooltipFormat) {
-				label = momentify(label, timeOpts).format(timeOpts.tooltipFormat);
+				return momentify(label, timeOpts).format(timeOpts.tooltipFormat);
 			}
 
-			return label;
+			// find the label with the most parts (milliseconds, minutes, etc.)
+			// format all labels with the same level of detail as the most specific label
+			// store this so that we only have to do it once
+			for (i = 0; i < len; i++) {
+				if (!data.labels || i > data.labels.length) {
+					continue;
+				}
+				momentDate = momentify(data.labels[i], timeOpts);
+				if (momentDate.millisecond() !== 0) {
+					me._labelFormatString = 'MMM D, YYYY h:mm:ss.SSS a';
+					break;
+				}
+			}
+			if (typeof me._labelFormatString === 'undefined') {
```

Comment:
I'm not sure to understand the caching logic: we iterate labels everytime (first loop) and if we don't find a time with millisecond at 0, we iterate again to find a time with second, minute and hour at 0, and if none, fallback to `'MMM D, YYYY'`. It doesn't seem to match the comment above: `store this so that we only have to do it once`.